src/main/java/seedu/mtracker/filemanager/InstrumentDecoder.java

```diff
@@ -0,0 +1,74 @@
+package seedu.mtracker.filemanager;
+
+import seedu.mtracker.model.Instrument;
+import seedu.mtracker.model.InstrumentManager;
+import seedu.mtracker.model.subinstrument.Crypto;
+import seedu.mtracker.model.subinstrument.Etf;
+import seedu.mtracker.model.subinstrument.Forex;
+import seedu.mtracker.model.subinstrument.Stock;
+
+import java.util.List;
+
+public class InstrumentDecoder {
+    public static void readFile(InstrumentManager instrumentManager, List<String> fileData) {
+        fileData.stream()
+                .forEach((line) -> {
+                    String[] textSegment = line.split(";", -1);
+                    switch (textSegment[0]) {
+                    case ("Crypto"):
+                        addCryptoToList(textSegment, instrumentManager);
+                        break;
+                    case ("Stock"):
+                        addStockToList(textSegment, instrumentManager);
+                        break;
+                    case ("Forex"):
+                        addForexToList(textSegment, instrumentManager);
+                        break;
+                    default:
+                        addEtfToList(textSegment, instrumentManager);
+                    }
+                });
+    }
+
+    public static void addCryptoToList(String[] textSegment, InstrumentManager instrumentManager) {
+        String name = textSegment[1];
+        double currentPrice = Double.parseDouble(textSegment[2]);
+        String sentiment = textSegment[3];
+        String expiry = textSegment[4];
+        String remarks = textSegment[5];
+        Instrument crypto = new Crypto(name, currentPrice, sentiment, expiry, remarks);
+        instrumentManager.addInstrument(crypto);
+    }
+
+    public static void addForexToList(String[] textSegment, InstrumentManager instrumentManager) {
+        String name = textSegment[1];
+        double currentPrice = Double.parseDouble(textSegment[2]);
+        String sentiment = textSegment[3];
```

Comment:
Looking this reminds me of the instrumentParser, do you think you can create the child classes like ForexDecoder etc each one add to list, it helps to reduce the repeating of general attributes

Reply:
Not necessary in this PR, in the interest of time. But if its something u think its good, can create a issue for it and put under the code quality issue